src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+
+@WebServlet("/chat")
+public class ChatServlet extends HttpServlet{
+    private static final String MESSAGE_TEXT_PROPERTY = "message-text";
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String MESSAGE_KIND = "Message";
+
+  /**
+   * When called, doGet will query all previously posted messages to the chat,
+   * and render the messages into a HTML template.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Calls query on all entities of type Message
+    Query messageQuery = new Query(MESSAGE_KIND).addSort(TIMESTAMP_PROPERTY, SortDirection.ASCENDING);
+    PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
+
+    // Creates list of the messages text
+    ImmutableList<String> messagesList = Streams.stream(preparedMessageQuery.asIterable()).map(message -> 
+        (String) message.getProperty(MESSAGE_TEXT_PROPERTY)).collect(toImmutableList());
+
+    // Data will be passed in as a list of messages in a map (needed for template)
+    ImmutableMap<String, ImmutableList<String>> data = ImmutableMap.of("messages", messagesList);
+
+    // File path starts in target/portfolio-1
+    SoyFileSet sfs = SoyFileSet
+        .builder()
+        .add(new File("../../src/main/java/templates/chat.soy"))
+        .build();
+    SoyTofu tofu = sfs.compileToTofu();
+
+    final String out = tofu.newRenderer("templates.chat.chatPage").setData(data).render();
+    response.getWriter().println(out);
+  }
+
+/**
+ * Is called by a button on the chat form. doPost takes the message and its
+ * attributes (such as timestamp and user) and stores them into Datastore.
+ */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+    final long timestamp = System.currentTimeMillis();
+    
+    // Reads API Key and Referer from file 
+    File apiKeyFile = new File("../../keys.txt");
+    Scanner scanner = new Scanner(apiKeyFile);
+    final String API_KEY = scanner.nextLine();
+    final String REFERER = scanner.nextLine();
+    scanner.close();
+
+    final String perspectiveURL = "https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=" + API_KEY;
+ 
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    HttpPost httpPost = new HttpPost(perspectiveURL);
+ 
+    PerspectiveRequest messageObject = new PerspectiveRequest(messageText);
+    final String inputJson = new Gson().toJson(messageObject);
+ 
+    httpPost.setEntity(new StringEntity(inputJson));
+    httpPost.addHeader("Referer", REFERER);
+    CloseableHttpResponse httpResponse = httpClient.execute(httpPost);
```

Comment:
same as the other comment, helper function along the lines of getHttpResponse() and move all the code in this line and above